### PR TITLE
add a Callable trait that is implemented for unsafe functions, too

### PR DESCRIPTION
**Pull request by oli-obk**
This trait will be used to mark functions as const evaluable (possibly by giving this trait a const generic bool parameter that is true for const and false otherwise)

I do not expect this trait to be ever visible to users, but to be a low level implementation detail

I have not moved `FnOnce::Output` to `Callable` yet as that is very invasive. I think we may want to delay that until the old solver is gone.

With appropriately improved (via `rustc_on_unimplemented`) error messages for when `Callable` doesn't hold, we could replace most, if not all, logic around number of argument mismatch and type of argument mismatch.

pulled out of https://github.com/rust-lang/rust/pull/101900

### CODE REVIEW

**Comment by ChayimFriedman2** [comment]
Is the log.log there on purpose?

**Comment by oli-obk** [comment]
yea ^^ for @lcnr's "convenience" as they wanted to see the current status

File: tests/ui/traits/new-solver/object-unsafety.stderr [hunk 1]
```diff
@@ -14,6 +14,12 @@ help: consider introducing a `where` clause, but there might be an alternative b
 LL | pub fn copy_any<T>(t: &T) -> T where dyn Setup<From = T>: Setup {
    |                                ++++++++++++++++++++++++++++++++
 
-error: aborting due to previous error
+error[E0277]: the trait bound `for<'a> fn(&'a <dyn Setup<From = T> as Setup>::From) -> <dyn Setup<From = T> as Setup>::From {copy::<dyn Setup<From = T>>}: Callable<(&T,)>` is not satisfied
+  --> $DIR/object-unsafety.rs:12:5
+   |
+LL |     copy::<dyn Setup<From = T>>(t)
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Callable<(&T,)>` is not implemented for fn item `for<'a> fn(&'a <dyn Setup<From = T> as Setup>::From) -> <dyn Setup<From = T> as Setup>::From {copy::<dyn Setup<From = T>>}`
```
Comment (oli-obk) [inline tests/ui/traits/new-solver/object-unsafety.stderr:21]: Not exactly sure what happened here, but I'm assuming we just don't explain that `Callable` isn't implemented because `Setup` isn't implemented for `dyn Setup`, but just fail the outer obligation.

Orthogonally to that I can make it less confusing by adding a `rustc_on_unimplemented` to `Callable` to give a more targeted message.

File: compiler/rustc_hir_typeck/src/callee.rs [hunk 2]
```diff
@@ -722,12 +731,57 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+    ) {
+        if value.references_error() {
+            return;
+        }
+
+        let context = self.tcx.hir().enclosing_body_owner(hir_id);
+        let kind = self.tcx.def_kind(context.to_def_id());
+        debug_assert_ne!(kind, DefKind::ConstParam);
+
+        let ty::FnDef(did, substs) = *value.kind() else {
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/callee.rs:751]: Why are we not also registering these predicates for other types?

File: compiler/rustc_hir_typeck/src/callee.rs [hunk 3]
```diff
@@ -722,12 +731,57 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+            trace!("not a function item");
+            return;
+        };
+        trace!(?substs);
+
+        // Tuple struct/variant constructors are not really function calls, they are always fine.
+        if let DefKind::Ctor(_, def::CtorKind::Fn) = self.tcx.def_kind(did) {
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/callee.rs:758]: Is this (and other early-returns below) for perf? Why not just register the predicate anyways even if we know it'll trivially hold? The solver(s) (ideally) should be able to handle a few extra trait preds.

File: compiler/rustc_hir_typeck/src/callee.rs [hunk 4]
```diff
@@ -722,12 +731,57 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         err.emit()
     }
 
+    /// Enforces that things being called actually are callable
+    #[instrument(skip(self, arguments))]
+    pub(super) fn check_callable(
+        &self,
+        hir_id: hir::HirId,
+        span: Span,
+        value: Ty<'tcx>,
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/callee.rs:740]: nit:

```suggestion
        callable_ty: Ty<'tcx>,
```

perhaps?

File: compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs [hunk 5]
```diff
@@ -92,18 +92,17 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         tuple_arguments: TupleArgumentsFlag,
         expected: Expectation<'tcx>,
     ) -> Ty<'tcx> {
-        let has_error = match method {
-            Ok(method) => method.substs.references_error() || method.sig.references_error(),
-            Err(_) => true,
-        };
-        if has_error {
+        let method = method.ok().filter(|method| !method.references_error());
+        let Some(method) = method else  {
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:96]:
```suggestion
        let Some(method) = method else {
```

File: compiler/rustc_hir_typeck/src/method/confirm.rs [hunk 6]
```diff
@@ -110,11 +112,11 @@ impl<'a, 'tcx> ConfirmContext<'a, 'tcx> {
         // could alter our Self-type, except for normalizing the receiver from the
         // signature (which is also done during probing).
         let method_sig_rcvr = self.normalize(self.span, method_sig.inputs()[0]);
-        debug!(
-            "confirm: self_ty={:?} method_sig_rcvr={:?} method_sig={:?} method_predicates={:?}",
-            self_ty, method_sig_rcvr, method_sig, method_predicates
-        );
+        debug!(?self_ty);
+        debug!(?method_sig_rcvr);
+        debug!(?pick);
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/method/confirm.rs:117]:
```suggestion
        debug!(?self_ty, ?method_sig_rcvr, ?pick);
```

File: compiler/rustc_infer/src/infer/opaque_types/table.rs [hunk 7]
```diff
@@ -40,7 +40,7 @@ impl<'tcx> OpaqueTypeStorage<'tcx> {
 
 impl<'tcx> Drop for OpaqueTypeStorage<'tcx> {
     fn drop(&mut self) {
-        if !self.opaque_types.is_empty() {
+        if !self.opaque_types.is_empty() && !std::thread::panicking() {
```
Comment (compiler-errors) [inline compiler/rustc_infer/src/infer/opaque_types/table.rs:43]: could you pull this out into another PR?

File: compiler/rustc_trait_selection/src/solve/trait_goals/structural_traits.rs [hunk 8]
```diff
@@ -182,18 +182,29 @@ pub(super) fn instantiate_constituent_tys_for_copy_clone_trait<'tcx>(
...
+    mode: CallableMode,
 ) -> Result<Option<ty::Binder<'tcx, (Ty<'tcx>, Ty<'tcx>)>>, NoSolution> {
+    use CallableMode::*;
     match *self_ty.kind() {
         // keep this in sync with assemble_fn_pointer_candidates until the old solver is removed.
         ty::FnDef(def_id, substs) => {
             let sig = tcx.fn_sig(def_id);
-            if sig.skip_binder().is_fn_trait_compatible()
-                && tcx.codegen_fn_attrs(def_id).target_features.is_empty()
+            if matches!(mode, Callable)
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/solve/trait_goals/structural_traits.rs:205]: I think it's easier to understand if we just match on `(*self_ty.kind(), mode)`

File: compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs [hunk 9]
```diff
@@ -320,6 +321,34 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
...
+        trace!(ty = ?self_ty.kind());
+        match *self_ty.kind() {
+            ty::Infer(ty::TyVar(_)) => {
+                debug!("assemble_builtin_callable_candidates: ambiguous self-type");
+                candidates.ambiguous = true; // Could wind up being a fn() type.
+            }
+            // Unsafe or extern functions do not necessarily implement FnOnce,
+            // but they do implement `Callable`
+            ty::FnDef(..) | ty::FnPtr(..) | ty::Closure(..) => {
+                candidates.vec.push(BuiltinCandidate { has_nested: false })
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs:344]: This incorrectly implements `fn(): Callable<(i32,)>` doesn't it?

File: compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs [hunk 10]
```diff
@@ -157,8 +160,8 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         c_variadic: bool,
         // Whether the arguments have been bundled in a tuple (ex: closures)
         tuple_arguments: TupleArgumentsFlag,
-        // The DefId for the function being called, for better error messages
-        fn_def_id: Option<DefId>,
+        // The callee type (e.g. function item ZST, function pointer or closure).
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:163]: It may be useful to note that this will be the `FnDef` for `FnOnce::call_once` if, for example, we try to call a `Box<dyn FnOnce()>`, etc.

File: compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs (hunk 9 above)
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs:344]:
```rust
#![feature(fn_traits)]

fn test<T: std::ops::Callable<()>>() {

}

fn main() {
    test::<fn(i32) -> i32>();
}
```

That passes on the current solver. See `confirm_fn_pointer_candidate` for how to extract the args tuple + `SelectionContext::confirm_poly_trait_refs` to confirm.

File: compiler/rustc_hir_typeck/src/callee.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_hir_typeck/src/callee.rs:751]: this is leftover cruft from the constness logic. We don't need it anymore now that we use the `Callable` trait. w

File: compiler/rustc_trait_selection/src/solve/trait_goals/structural_traits.rs (hunk 8 above)
Comment (oli-obk) [reply compiler/rustc_trait_selection/src/solve/trait_goals/structural_traits.rs]: I did that for the closures, but I don't see how we can improve anything for `FnDef` or `FnPtr`.

File: tests/ui/chalkify/assert.rs [hunk 11]
```diff
@@ -1,6 +0,0 @@
-// run-pass
```
Comment (compiler-errors) [inline tests/ui/chalkify/assert.rs:1]: lol where did these tests go

File: tests/ui/traits/new-solver/object-unsafety.stderr (hunk 1 above)
Comment (compiler-errors) [reply tests/ui/traits/new-solver/object-unsafety.stderr:21]: It's because `&'a <dyn Setup<From = T> as Setup>::From` isn't normalizing to `&T`.

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
This looks fine. I'm tempted to run a perf run and crater on this, just because the fn code in the old solver is so delicate and I'd rather not have to revert this during the beta check. Sorry for the delay in reviewing, and for the additional delay in waiting for crater. Thanks for working on this :pray: 

@bors try @rust-timer queue

**Comment by compiler-errors** [comment]
Oh yeah crater

@craterbot check

**Comment by compiler-errors** [comment]
I'll do triage on that crater report in the next few days.

**Comment by oli-obk** [comment]
Most of these are breakages in crates that already require nightly.

But there are some like

* https://crater-reports.s3.amazonaws.com/pr-107123/try%23b5a2ddda2086ab9fec9b967656c474e1f7b38e32/reg/linfa-logistic-0.6.1/log.txt
* https://crater-reports.s3.amazonaws.com/pr-107123/try%23b5a2ddda2086ab9fec9b967656c474e1f7b38e32/reg/ontv-0.1.3/log.txt
* https://crater-reports.s3.amazonaws.com/pr-107123/try%23b5a2ddda2086ab9fec9b967656c474e1f7b38e32/reg/uom-0.34.0/log.txt

that are a regression to stable. I still need to build MCVEs, but the issue appears to be around method lookup

File: tests/ui/chalkify/assert.rs (hunk 11 above)
Comment (oli-obk) [reply tests/ui/chalkify/assert.rs:1]: Fixing them would have required implementing `Callable` in chalk

**Comment by oli-obk** [comment]
@bors try @rust-timer queue

I checked a few of the failing crates locally, after I rebased they stopped failing.

I'm gonna rerun crater on just the failed crates

**Comment by oli-obk** [comment]
@craterbot check crates=list:https://crater-reports.s3.amazonaws.com/pr-107123/retry-regressed-list.txt

**Comment by oli-obk** [comment]
@craterbot check crates=https://crater-reports.s3.amazonaws.com/pr-107123/retry-regressed-list.txt

**Comment by JohnCSimon** [comment]
@oli-obk 

ping from triage - can you post your status on this PR? This PR has not received an update in a few months. Thank you!

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks
